File: src/main/java/de/hsa/bpmnandai/project/domain/facade/ExternalDmnFacade.java
package de.hsa.bpmnandai.project.domain.facade;

import de.hsa.bpmnandai.project.domain.service.CreditService;
import de.hsa.bpmnandai.project.domain.service.DroolsService;
import de.hsa.bpmnandai.project.domain.service.ProcessService;
import de.hsa.bpmnandai.project.infrastructure.entity.BankLoanRequestEntity;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;

@Slf4j
@Component
@RequiredArgsConstructor
public class ExternalDmnFacade {

    private final ProcessService processService;
    private final CreditService creditService;
    private final DroolsService droolsService;

    private final String PROCESS_KEY = "Process_KreditAnfrage_Ext";

    public void processLoanRequests() {
        final List<BankLoanRequestEntity> allRequests = creditService.getAllRequests();
        int iterationCounter = 1;
        for (final BankLoanRequestEntity request : allRequests) {
            final String businessKey = request.getId();
            final Map<String, Object> vars = Map.of("requestId", businessKey);
            log.info("Starting instance '{}' of '{}' with key '{}'", iterationCounter++, allRequests.size(), businessKey);
            processService.startInstanceOfProcess(PROCESS_KEY, businessKey, vars);
        }
    }

    public String checkCreditworthiness(final String requestId) {
        final BankLoanRequestEntity bankLoanRequest = this.creditService.getSpecificRequest(requestId);
        final String result = droolsService.evaluateCreditApprovalDecisionModel(bankLoanRequest);
        bankLoanRequest.setPredictionIsApproved2(result);
        creditService.saveCreditRequest(bankLoanRequest);
        return result;
    }

}

File: src/main/kotlin/de/emaarco/bpmnandai/credit/domain/facade/ExternalDmnFacade.kt
package de.emaarco.bpmnandai.credit.domain.facade

import de.emaarco.bpmnandai.camunda.domain.ProcessService
import de.emaarco.bpmnandai.credit.domain.service.CreditService
import de.emaarco.bpmnandai.credit.domain.service.DroolsService
import de.emaarco.bpmnandai.credit.infrastructure.entity.CreditRequestEntity
import mu.KotlinLogging
import org.springframework.stereotype.Component

@Component
class ExternalDmnFacade(
    private val processService: ProcessService,
    private val creditService: CreditService,
    private val droolsService: DroolsService,
) {

    private val log = KotlinLogging.logger {}
    private val processKey = "Process_KreditAnfrage_Ext"

    fun processLoanRequests() {
        val allRequests: List<CreditRequestEntity> = creditService.getAllRequests()
        var iterationCounter = 1
        for (request in allRequests) {
            val businessKey: String = request.requestId
            val vars: MutableMap<String, Any> = HashMap()
            vars["requestId"] = businessKey
            log.info("Starting instance '${iterationCounter++}' of '${allRequests.size}' with key '${businessKey}'")
            processService.startInstanceOfProcess(processKey, businessKey, vars)
        }
    }

    fun checkCreditworthiness(requestId: String): String {
        val bankLoanRequest: CreditRequestEntity = creditService.getSpecificRequest(requestId)
        val result: String = droolsService.evaluateCreditApprovalDecisionModel(bankLoanRequest)
        bankLoanRequest.predictionIsApproved2 = result
        creditService.saveCreditRequest(bankLoanRequest)
        return result
    }
}